cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactInspector.java

```diff
@@ -188,6 +201,81 @@ public ArtifactClasses inspectArtifact(Id.Artifact artifactId, File artifactFile
   }
 
   /**
+   * we instantiate the plugin and invoke the plugin,
+   * we deserialize the request body into parameter type and pass it when invoking the plugin method and also
+   * optionally pass {@link EndpointPluginContext} to the method if its declared in the plugin method.
+   * @param pluginInstantiator to instantiate plugin instances.
+   * @param pluginClass call having the plugin method to invoke
+   * @param methodParameter parameter body
+   * @param endpointPluginContext endpoint plugin context that can optionally be passed to the method
+   * @throws IOException if there was an exception opening the jar file
+   * @throws InvalidArtifactException if the artifact is invalid
+   * @throws NotFoundException Not Found exception thrown from plugin method invocation or if no matching plugin found.
+   * @throws InvocationTargetException exception thrown from plugin method invocation.
+   */
+  public String invokePluginEndpoint(PluginInstantiator pluginInstantiator, Id.Artifact artifact,
+                                     PluginClass pluginClass, String methodName, String methodParameter,
+                                     DefaultEndpointPluginContext endpointPluginContext)
+    throws IOException, BadRequestException, NotFoundException, ClassNotFoundException, InvocationTargetException {
+
+    // See if there are export packages. Plugins should be in those packages
+
+    ClassLoader pluginClassLoader = pluginInstantiator.getArtifactClassLoader(artifact.toArtifactId());
+
+    Class pluginClassLoaded = pluginClassLoader.loadClass(pluginClass.getClassName());
+
+    Method[] methods = pluginClassLoaded.getDeclaredMethods();
```

Comment:
Also, we are doing this for every endpoint call? Shouldn't we preserve the information when we inspect a plugin at deployment time?